### PR TITLE
added tests for Create and Split Cascade Refactoring

### LINKED ISSUE #12533: Add tests for `RBCreateCascasde` and `RBSplitCascade`
`RBCreateCascasde` and `RBSplitCascade` have no test, in order to expose them we need to be certain that they work.

### PR DESCRIPTION
**Pull request by jainoshika**
Tests are added for `RBCreateCascadeRefactoring` and `RBSplitCascadeRefactoring` in Refactoring-Transformations-Tests. 

To broaden test coverage, some edge cases are explored with the help of an LLM. These tests were reviewed to ensure they are correct for the refactoring. If any changes do let me know.

Fixes #12533

### CODE REVIEW

File: src/Refactoring-Transformations-Tests/RBCreateCascadeRefactoringTest.class.st [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	#category : 'Refactoring-Transformations-Tests-Test',
+	#package : 'Refactoring-Transformations-Tests',
+	#tag : 'Test'
+}
+
+{ #category : 'utilities' }
+RBCreateCascadeRefactoringTest >> createCascadeFrom: sourceToCombine in: source [
+
+| method start |
+
```
Comment (Ducasse) [inline src/Refactoring-Transformations-Tests/RBCreateCascadeRefactoringTest.class.st]: please format the code

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ref generateChanges.
+self assert:
+    (class parseTreeForSelector: #example)
+    equals:
+    (self parseMethod: 'example
+    obj foo.
+    x := obj
+        bar;
+        baz.')
+]
```
Comment (Ducasse) [inline src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:69]: the previous example is not really good. 
I would expect 

```
obj foo.
obj bar.
x := obj baz
```

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+"Step 4: compare ASTs"
+self assert:
+    (class parseTreeForSelector: #example)
+    equals: originalMethod
+]
+
+{ #category : 'tests' }
+RBSplitCascadeRefactoringTest >> testExtractReceiver [
+
```
Comment (Ducasse) [inline src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st]: This test is not worth

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	#name : 'RBSplitCascadeRefactoringTest',
+	#superclass : 'RBAbstractTransformationTest',
+	#instVars : [
+		'class'
+	],
+	#category : 'Refactoring-Transformations-Tests-Test',
+	#package : 'Refactoring-Transformations-Tests',
+	#tag : 'Test'
+}
+
```
Comment (Ducasse) [inline src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:11]: I have the impression that the from: argument is not that great

Comment (jainoshika) [reply src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:11]: i changed it to `#example` currently. if you may share the reasons behind, i will try to think of more.

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st (hunk 2 above)
Comment (jainoshika) [reply src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:69]: yes, that should be the expected behaviour. but i see the method implementation does not support complete flattening and assigning the last, rather it depends on where the cursor is placed.

**Comment by Ducasse** [comment]
I have the impression that this refactoring is not fully specified.

**Comment by jainoshika** [comment]
> I have the impression that this refactoring is not fully specified.

I agree, shall i go through the implementation first and  check out the issues in it?

**Comment by jainoshika** [comment]
I made changes in the `splitCascade` which now flattens cascade and returns last in Assignment. 

But i am quite confused how the refactoring is supposed to work. the comment says that it divides in 2 parts so then, 
will obj `a; b; ^ c; d.` return `obj a;b. `and `obj c; d`
But in case of assignment, we want the method to flatten the whole cascade like 
```
obj a. 
obj b. 
obj c. 
x = obj d.
```

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	| ref |
+	ref := self
+    	splitCascadeFrom: 'foo'
+    	in: 'example
+    	obj foo;
+        	bar;
+        	baz'.
+	self should: [ ref generateChanges ]
+    	raise: RBRefactoringError.
+]
```
Comment (balsa-sarenac) [inline src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:127]: For me this would pass, I'd split after first message here, better UX too.

I'm not saying we should implement this in this PR.

File: src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	| ref |
+	ref := self
+    	splitCascadeFrom: 'foo; bar; baz'
+    	in: 'example
+    	obj foo;
+        	bar;
+        	baz'.
+	self should: [ ref generateChanges ]
+    	raise: RBRefactoringError.
+]
```
Comment (balsa-sarenac) [inline src/Refactoring-Transformations-Tests/RBSplitCascadeRefactoringTest.class.st:193]: I'd also expect this to do:
```smalltalk
obj foo.
obj bar.
obj baz.
```

But this might only be me.
I'm not saying we should do this in this PR, we should discuss this.

**Review by balsa-sarenac — COMMENTED** [review]
I think it's really good that we got these tests!
What I don't like is how split cascade works right now. For me, it has weird interface. We should discuss what would be a good interface.

I'd like to suggest we create a small UI that by default does best guess split, then you have small check boxes between message calls where you can check them to make those calls be cascade or not, so user ultimately picks what will be split and what isn't.

We can also rely on selection for splitting - example: all cascades in selection will be split, before or after it we preserve cascade. This might be harder since users don't know this and is not intuitive.

**Comment by jainoshika** [comment]
> I'd like to suggest we create a small UI that by default does best guess split, then you have small check boxes between message calls where you can check them to make those calls be cascade or not, so user ultimately picks what will be split and what isn't.

I really like this idea, it will give the user more options and control.
